keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,203 @@
+is often required as a Parameter can change a value in a ConfigMap that is used by the Pods. If the Pods are not restarted
+the change is propagated to the Pods, but as most applications are not aware of changing config files they require a 
+restart. This was originally solved by using [stakater/Reloader](https://github.com/stakater/Reloader) but this had 
+[issues](https://github.com/kudobuilder/kudo/issues/1025) which lead to the self developed solution, which is at the 
+moment applied to [all templates](https://github.com/kudobuilder/kudo/issues/1036)
+
+This behavior can currently not be controlled and is always active. The issue is that not all Parameter changes require
+a rolling restart of Pods, and sometimes the restart is unwanted and can negatively affect the state of the application
+the operator is controlling.
+
+One example would be the `replica` count of a StatefulSet: An increase here should only start new Pods and not restart
+all existing ones. Another example would be an application that regularly re-reads the config files, in this case the 
+pod doesn't need to be restarted if a config map changes - the application would pick up changed values automatically.
+
+A new flag introduced by this KEP will allow an operator developer to control when a changed parameter will trigger a 
+restart of pods from a deployment or stateful set.
+
+### Goals
+
+Make it possible for an operator developer to specify when a parameter will restart all pods from a stateful set or a deployment.
+
+### Non-Goals
+
+- Preventing Pod restart where plain k8s would do it: If the spec template for a stateful set is changed, Kubernetes will still
+restart all pods automatically - there is no way to prevent that.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+- Restarting a stuck Pod
+
+## Proposal
+
+Allow the operator developer to define groups of parameters that can then be used to trigger a restart of pods from
+a deployment or stateful set.
+
+Add a list of `groups` that each parameter is part of:
+
+```yaml
+  - name: CONFIG_VALUE
+    description: "A parameter that changes a config map which is used by a stateful set."
+    default: "3"
+    groups:
+      - MAIN_GROUP
+  - name: SOME_OTHER_PARAMETER
+    description: "A parameter that changes something that is used by two stateful sets."
+    default: "3"
+    groups:
+      - MAIN_GROUP
+      - ADDITIONAL_GROUP
+```
+
+KUDO then calculates the hash of all parameters in a group. This hash can then be used
+instead of the the `last-plan-execution-uid` in a deployment or stateful set to trigger the reload of the pods:
+
+```yaml
+spec:
```

Comment:
lets take the example of `Cassandra` so if a statefulset is using `MAIN_GROUP` like this:
```
annotations:
        config-hash: {{ .ParamGroups.MAIN_GROUP.hash }}
```
if the a `configmap` uses replica to generate the configuration as Cassandra does
```
seeds: "{{- range $i, $node := until (int (min 2 .Params.NODE_COUNT)) -}}
```
that would still trigger a rolling restart of the stateful set with this solution. Even the content of `seeds` is the same in case of scaling up, but might be different in case of scaling down. 

So we will be asking operator developers to restrict the use of the parameters to generate configuration they need?

Reply:
That would depend on if the `NODE_COUNT` param would be in the `MAIN_GROUP`. 
In the case of Cassandra, I would argue that `NODE_COUNT` would not be in the `MAIN_GROUP`, even though it is used in the specific config map.

Which is actually an interesting argument against the automatic dependency graph...

Reply:
>  I would argue that NODE_COUNT would not be in the MAIN_GROUP, even though it is used in the specific config map.

that's buggy, as what if seeds content has actually changed and we do really need a rolling restart?
> Which is actually an interesting argument against the automatic dependency graph...

I see its an argument against the current approach 🤔 
if the dependency graph is implemented correctly it will know that seeds are same so no need to restart, and if they aren't the same then it totally valid to restart all pods.

Reply:
I guess that depends on how the dependency graph is build. At the moment, the idea is to determine which parameters are used in which resource - and if any parameter changed, we need to restart.

Maybe I'd be possible to fully ignore the parameters in the dependency graph and only look at the rendered resources: If the rendered resource is the same as the currently deployed one, we don't need to restart.

The problem with this is the comparison, as K8s adds all the additional fields. But we do something similar with the 3 way merge when applying resources, so it might be possible.